Issue #7277 - Allow `Request.getLocalName()` and `.getLocalPort()` to be overridden (Split authority approach)

joakime opened the pull request:
This is the split authority approach as proposed by @sbordet as an alternate approach for PR #7283

joakime committed:
Issue #7277 - ServletRequest.getLocalName() override

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Allow Request.getLocalName() to be overridden

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Updating based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - More updates based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Switch from setLocalName to setLocalAuthority per review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Fixing NPE from updated getLocalPort

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Updates based on review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Split Authority Approach

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -161,7 +163,18 @@
+     * The authority used to define the server uri authority fallback for the connection
+     * (see ServletRequest.getServerName(), and ServletRequest.getServerPort()).
+     */
+    private HostPort _serverUriAuthority;
```
I think just `_serverAuthority` is a better name.

joakime replied:
Done.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -298,6 +311,43 @@ public int getAcceptors()
+    }
+
+    @ManagedAttribute("server uri authority")
+    public HostPort getServerUriAuthority()
```
```suggestion
    public HostPort getServerAuthority()
```

joakime replied:
Done.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -161,7 +163,18 @@
+     * The authority used to define the local authority for the connection
+     * (see ServletRequest.getLocalName(), ServletRequest.getLocalAddr(), and ServletRequest.getLocalPort()).
+     */
+    private HostPort _localAuthority;
```
is this really a local authority?   Feels more like a local address

joakime replied:
Changed to SocketAddress

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,6 +298,64 @@ public EndPoint getEndPoint()
+     * null in the case of no local address (usually seen in connectors not based on IP networking).
+
+     */
+    public String getLocalName()
```
Why convert to separate name/port methods at the channel level?
I think it would be better if the channel provided `getServerAuthority()` and `getLocalAuthority()` (or `getLocalAddress()` if name change above is accepted).    We only need separate name/port methods in the request where they are defined by servlet API

gregw replied:
Note that the only real reason I see to resolve to name/port methods here is that the existing `getLocalAddress()` returns an `InetSocketAddress`, whilst the new methods return `HostPort`.      I'm thinking it would be simpler to just use `InetSocketAddress` everywhere.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1428,20 +1409,24 @@ private String findServerName()
             }
```
Do we have coverage for the code above  here?   I.e. does the `findServerName()` method ever find a `Host` header field? as I'm pretty sure that will have already been copied to the authority in the HttpURI by now.... perhaps if a customizer added a Host header but didn't update the HttpURI?

joakime replied:
Yep, this block is used if a `Customizer` sets a `Host` header without updating the authority.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1428,20 +1409,24 @@ private String findServerName()
 
+        if (_channel != null)
+        {
+            Connector connector = _channel.getConnector();
+            if (connector != null)
+            {
+                HostPort serverAuth = connector.getServerUriAuthority();
+                if (serverAuth != null)
+                {
+                    return formatAddrOrHost(serverAuth.getHost());
+                }
+            }
```
This feels wrong to be diving down to the connector  here.  Would be better just to ask for `_channel.getServerAuthority()`

joakime committed:
Issue #7277 - Fixing null return in Request.getServerName() usage

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Addressing review comments

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - SocketAddress as connector.getLocalAddress

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Correcting field name

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Fixing MavenServerConnector compilation issue

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -161,7 +164,18 @@
     private long _idleTimeout = 30000;
     private String _defaultProtocol;
     private ConnectionFactory _defaultConnectionFactory;
+    /* The name used to link up virtual host configuration to named connectors */
     private String _name;
+    /**
+     * The authority used to define the server uri authority fallback for the connection
+     * (see ServletRequest.getServerName(), and ServletRequest.getServerPort()).
+     */
+    private HostPort _serverAuthority;
+    /**
+     * The address used to define the local address for the connection
+     * (see ServletRequest.getLocalName(), ServletRequest.getLocalAddr(), and ServletRequest.getLocalPort()).
+     */
+    private SocketAddress _localAddress;
```
I think it is sufficient to have javadoc on the associated getters/setters rather than on the private fields

joakime replied:
I'll move these to the `Connector` apidoc, as it's kinda useful information not represented there currently.

joakime replied:
Moved

gregw commented on jetty-maven-plugin/src/main/java/org/eclipse/jetty/maven/plugin/MavenServerConnector.java:
```diff
@@ -55,6 +58,8 @@
     private String name;
     private int port;
     private long idleTimeout;
+    private HostPort serverAuthority;
+    private SocketAddress localAddress;
```
Can't you just use these from the delegate connector?

joakime replied:
delegate connector?  Let me review that code again, I didn't notice that.

joakime replied:
Fixed

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java:
```diff
@@ -283,6 +285,19 @@ public void setInheritChannel(boolean inheritChannel)
 
+    private void setAcceptChannel(ServerSocketChannel acceptChannel) throws IOException
+    {
+        _acceptChannel = acceptChannel;
+        _localPort = _acceptChannel.socket().getLocalPort();
+        SocketAddress localAddress = _acceptChannel.socket().getLocalSocketAddress();
+        if (localAddress instanceof InetSocketAddress)
+            _localName = ((InetSocketAddress)localAddress).getHostString();
+        else
+            _localName = localAddress.toString();
+        if (_localPort <= 0)
+            throw new IOException("Server channel not bound");
```
It is kind of strange that the `localAddress` here is used to initialize the `_localName`, but not the `_localAddress` or `_serverAuthority`.

I know you said there is a reason??? was it the Proxy protocol?   So either we need to make another attempt to initialize these here or we have to document somewhere in the code why not.   If they are not going to be initialized, then we need better javadoc on getLocalAddress to say what null return means

joakime replied:
I reverted this.  It's left over from when we attempted to default the localAddress during doStart().
That (the defaulting of localAddress at doStart) couldn't be done for the proxy cases.

gregw reviewed (changes requested):
Getting close.   Mostly just needs some javadoc cleanups.
However, I would like to have one more go at always initializing the _localAddress and if that can't be done, document why

joakime committed:
Issue #7277 - Adding test cases for abs-uri

+ Also fixing left over names
  *.local.authority to *.local.name to
  represent the change in Connector.localAddress

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Use delegate in MavenServerConnector

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Remove redundant javadoc

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Restore AbstractNetworkConnector to original

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Improve APIdoc

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - revert ServerConnector.localName changes (no longer needed)

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Connector.java:
```diff
@@ -102,4 +104,48 @@
+    /**
+     * Get the optional Server URI authority default
+     *
+     * @return Returns the connection server authority (name/port).
```
```suggestion
     * @return Returns the connection server authority (name/port) or null.
```

I really don't like this "or null".

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/Connector.java:
```diff
@@ -102,4 +104,48 @@
+    /**
+     * Get the Local Address of the connection
+     *
+     * @return Returns the connection local address.
```
```suggestion
     * @return Returns the connection local address or null.
```
I really don't like the or null.

gregw reviewed (changes requested):
I really don't like the fact that in most situations `getLocalAddress` and `getServerAuthority` are going to return null.   If I see those methods on the connector, they should provide valid values if the connector is running.

Currently the semantic of the methods is `getLocalAddressDefault()` and `getServerAuthorityDefault()` or something like that... but "Default" is not the right word, nor is "Override", as any verb we put here is describing an algorithm that is implemented outside of this instance and is just assuming how these methods are used.     So it reeks of bad design the fact that we can't get the name right.   Currently these getters/setters are entirely unrelated to the behaviour of the connector and are essentially fields for an external algorithm.

Can you describe again the use-case that prevents us making these real methods that actually returns a dynamically determined local address and authority (if none have been set)?

joakime committed:
Issue #7277 - Using HttpConfiguration for configuration

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #7277 - Removing dead code

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -78,6 +81,9 @@
     private boolean _notifyRemoteAsyncErrors = true;
     private boolean _relativeRedirectAllowed;
 
+    private HostPort _serverAuthority;
+    private SocketAddress _localAddress;
```
We need to either add these fields to the copy constructor OR well document that they are not copied by the copy constructor.

I'm probably happy with either approach... whatever works best in the XML and makes more sense for configuring http and https connectors.

sbordet replied:
I would go for full copy of these fields, like all the others.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -662,6 +668,71 @@ public boolean isRelativeRedirectAllowed()
         return _relativeRedirectAllowed;
     }
 
+    /**
+     * Get the Local Address of the connection
+     *
+     * @return Returns the connection local address.
+     */
+    @ManagedAttribute("Local address override")
+    public SocketAddress getLocalAddress()
```
```suggestion
    /**
     * Get the SocketAddress override to be reported as the local address of all connection
     *
     * @return Returns the connection local address override or null.
     */
    @ManagedAttribute("Local address override")
    public SocketAddress getLocalAddress()
```

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -662,6 +668,71 @@ public boolean isRelativeRedirectAllowed()
+    }
+
+    /**
+     * Get the optional Server URI authority default
+     *
+     * @return Returns the connection server authority (name/port).
+     */
+    @ManagedAttribute("Server authority override")
+    public HostPort getServerAuthority()
```
```suggestion
     * Get the Server URI authority default to be used if no authority is provided by a request.
     *
     * @return Returns the connection server authority (name/port) or null
     */
    @ManagedAttribute("Server authority default")
    public HostPort getServerAuthority()
```

sbordet replied:
It's not the _URI_ authority, just the `server authority`.
The description of `@ManagedAttribute` should be "The server authority if none provided by requests".

gregw reviewed (dismissed):
Much better.
Just a couple of javadoc tweaks and the config copy constructor to resolve (one way or the other). 

Actually we should probably also define some properties in the mod files and update the xml that sets up our HttpConfiguration to test that it is usable as we think it should be.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +299,73 @@ public EndPoint getEndPoint()
+     * This is the name after the connector is bound and the connection is accepted.
+     * The value is typically the local address and optionally the local host name
+     * if the local address is unavailable.
+     * </p>
```
This javadoc paragraph is incorrect, the address is never returned, only the host name.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -662,6 +668,71 @@ public boolean isRelativeRedirectAllowed()
+
+    /**
+     * <p>
+     * Specify the connection server uri authority (name/port) used within application API layer
```
Here again, remove "uri" and add that's used if no authority is provided by the request.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -662,6 +668,71 @@ public boolean isRelativeRedirectAllowed()
+        if (authority == null)
+            _serverAuthority = null;
+        else if (!authority.hasHost())
+            throw new IllegalStateException("Server URI Authority must have host declared");
```
Remove "URI".

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:
```diff
@@ -662,6 +668,71 @@ public boolean isRelativeRedirectAllowed()
+     */
+    public void setLocalAddress(SocketAddress localAddress)
+    {
+        Objects.requireNonNull(localAddress, "Local Address");
+
+        _localAddress = localAddress;
```
```suggestion
        _localAddress = Objects.requireNonNull(localAddress);
```

sbordet replied:
Why can't it be set to `null`? I think it should, as the server authority can.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1428,21 +1409,25 @@ private String findServerName()
+                    return formatAddrOrHost(serverAuth.getHost());
+                }
+            }
+        }
```
This block should be moved to `HttpChannel.getServerAuthority()`, to make it similar to `HttpChannel.getLocalName()`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1474,19 +1460,29 @@ private int findServerPort()
+                if (serverAuth != null)
+                    return serverAuth.getPort();
+            }
+        }
```
Move this block to `HttpChannel.getServerAuthority()`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +299,73 @@ public EndPoint getEndPoint()
+     * scenarios where Jetty is being an intermediary and the local name
+     * needs to be changed to satisfy public (pre-intermediary) HTTP behaviors
+     * such as absolute-URI creation (eg: Location response header).
+     * </p>
```
This paragraph is incorrect. It's not used for URI creation (and should not be).

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +299,73 @@ public EndPoint getEndPoint()
+     * scenarios where Jetty is being an intermediary and the local port
+     * needs to be changed to satisfy public (pre-intermediary) HTTP behaviors
+     * such as absolute-URI creation (eg: Location response header).
+     * </p>
```
This paragraph is incorrect. It's not used for URI creation (and should not be).

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +299,73 @@ public EndPoint getEndPoint()
+     * such as absolute-URI creation (eg: Location response header).
+     * </p>
+     *
+     * @return the local authority port if overridden, or the local address port, or
```
Remove the part on "authority".

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +299,73 @@ public EndPoint getEndPoint()
+    public String getLocalName()
+    {
+        SocketAddress localAddress = getHttpConfiguration().getLocalAddress();
+        if (localAddress != null)
```
Null check can be removed, as the `instanceof` below would take care of it implicitly.

joakime committed:
Issue #7277 - Addressing review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -298,8 +300,89 @@ public EndPoint getEndPoint()
+
     public InetSocketAddress getLocalAddress()
     {
+        SocketAddress localAddress = getHttpConfiguration().getLocalAddress();
```
Null check for HttpConfiguration?

joakime committed:
Issue #7277 - Adding missing null check

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
